### PR TITLE
Rule to ensure test tasks use trusted artifacts

### PR DESCRIPTION
**Pull request by joejstuart**
https://issues.redhat.com/browse/EC-696

### CODE REVIEW

File: policy/release/trusted_task.rego [hunk 1]
```diff
@@ -91,6 +91,33 @@ deny contains result if {
...
+#     trusted. Note that trust is eventually revoked from Tasks when newer versions are made
+#     available.
+#   collections:
+#   - redhat
+#   effective_on: 2024-07-12T00:00:00Z
+#
+deny contains result if {
+	_uses_trusted_artifacts
+	some attestation in lib.pipelinerun_attestations
+	some test_task in tkn.tasks_test_output(attestation)
```
Comment (lcarva) [inline policy/release/trusted_task.rego:115]: I don't think we want to add a new set of rules here. We probably just want to rework the existing rules to pick up the builder and the test tasks.

Comment (joejstuart) [reply policy/release/trusted_task.rego]: https://github.com/enterprise-contract/ec-policies/pull/1040/commits/6bbee2166833b472245d8b76577ec175decd853e

File: policy/release/trusted_task.rego [hunk 2]
```diff
@@ -182,9 +182,10 @@ _trust_errors contains error if {
 	_uses_trusted_artifacts
 	some attestation in lib.pipelinerun_attestations
 	build_tasks := tkn.build_tasks(attestation)
-	some build_task in build_tasks
+	test_tasks := tkn.tasks_test_output(attestation)
+	some trust_task in array.concat(build_tasks, test_tasks)
```
Comment (simonbaird) [inline policy/release/trusted_task.rego:186]: nitpick: I don't love the name, maybe:

```suggestion
	some build_or_test_task in array.concat(build_tasks, test_tasks)
```

or...

```suggestion
	some significant_task in array.concat(build_tasks, test_tasks)
```

Comment (simonbaird) [reply policy/release/trusted_task.rego:186]:
```suggestion
	some needs_trust_task in array.concat(build_tasks, test_tasks)
```

idk..

**Review by simonbaird — APPROVED** [review]

File: policy/lib/tekton/task.rego [hunk 3]
```diff
@@ -173,6 +173,13 @@ build_tasks(attestation) := [task |
 	count(image_digest) > 0
 ]
 
+# return the tasks that have "TEST_OUTPUT" as a result
+tasks_test_output(attestation) := [task |
```
Comment (lcarva) [inline policy/lib/tekton/task.rego:177]: NIT: `test_tasks` ? We don't need to encode what it means to be a test task in the name.

File: policy/release/trusted_task_test.rego [hunk 4]
```diff
@@ -435,6 +456,22 @@ task_c := {
...
+	"invocation": {"parameters": {"B_ARTIFACT": artifact_b, "D_ARTIFACT": artifact_d}},
+	"results": [{
+		"name": "TEST_OUTPUT",
+		"value": "{\"FAILED\": \"1\"}",
+		"type": "string",
+	}],
+	"ref": {"name": "TaskA", "kind": "Task", "bundle": trusted_bundle},
+}
+
+attestation_test_ta := {"statement": {"predicate": {
```
Comment (lcarva) [inline policy/release/trusted_task_test.rego:470]: We could use the existing `attestation_ta` statement and just add `task_test_a` to it? (You'd have to change the value of `tekton.dev/pipelineTask` to be unique (and also probably the `.ref.name` value)

**Review by lcarva — APPROVED** [review]
Very nice!

**Comment by lcarva** [comment]
I did some testing to verify things are working properly. 

I created the following policy config:
```yaml
---
publicKey: |
  -----BEGIN PUBLIC KEY-----
  MFkwEwYHKoZIzj0CAQYIKoZIzj0DAQcDQgAEZP/0htjhVt2y0ohjgtIIgICOtQtA
  naYJRuLprwIv6FDhZ5yFjYUEtsmoNcW7rx2KM6FOXGsCX3BNc7qhHELT+g==
  -----END PUBLIC KEY-----
sources:
  - name: Release Policies
    data:
      - github.com/release-engineering/rhtap-ec-policy//data
      - /tmp/nervous-bouman/data
    policy:
      - /home/lucarval/src/enterprise-contract/ec-policies/policy/lib
      - /home/lucarval/src/enterprise-contract/ec-policies/policy/release
    config:
      include:
        - trusted_task
```

That points to my local system for the policy rules. It also points to my local version of the trusted tasks data. I modified the data so the `git-clone-oci-ta` Task is not trusted.

Without the changes from this PR, I get two violations:

```bash
🐚 ec validate image --ignore-rekor --policy policy.yaml --image quay.io/redhat-user-workloads/rhtap-contract-tenant/ec-v04/cli-v04@sha256:baa0e24c9faa7b0b8875daacbeff6587c71d450d7ad1b4f97f033e6a388a4e58 --output yaml | yq '.components[].violations[].msg'
Error: success criteria not met
Code tampering detected, untrusted PipelineTask "clone-repository" (Task "git-clone-oci-ta") was included in build chain comprised of: build-container, clone-repository, prefetch-dependencies
Code tampering detected, untrusted PipelineTask "clone-repository" (Task "git-clone-oci-ta") was included in build chain comprised of: build-source-image, clone-repository, prefetch-dependencies
```

With the changes from this PR, I get the expected three:

```
🐚 ec validate image --ignore-rekor --policy policy.yaml --image quay.io/redhat-user-workloads/rhtap-contract-tenant/ec-v04/cli-v04@sha256:baa0e24c9faa7b0b8875daacbeff6587c71d450d7ad1b4f97f033e6a388a4e58 --output yaml | yq '.components[].violations[].msg'
Error: success criteria not met
Code tampering detected, untrusted PipelineTask "clone-repository" (Task "git-clone-oci-ta") was included in build chain comprised of: build-container, clone-repository, prefetch-dependencies
Code tampering detected, untrusted PipelineTask "clone-repository" (Task "git-clone-oci-ta") was included in build chain comprised of: build-source-image, clone-repository, prefetch-dependencies
Code tampering detected, untrusted PipelineTask "clone-repository" (Task "git-clone-oci-ta") was included in build chain comprised of: clone-repository, prefetch-dependencies, sast-snyk-check
```

Good job! :taco:

File: policy/release/trusted_task_test.rego (hunk 4 above)
Comment (joejstuart) [reply policy/release/trusted_task_test.rego]: Good idea. That better reflects the attestations we work with.

**Pull request merged**